src/section/section_keys.rs

```diff
@@ -50,12 +65,92 @@ impl SectionKeysProvider {
 
     pub fn finalise_dkg(&mut self, public_key: &bls::PublicKey) {
         if let Some(share) = &self.pending {
-            let pending_public_key = share.public_key_set.public_key();
+            if *public_key != share.public_key_set.public_key() {
+                return;
+            }
+        }
+        if let Some(share) = self.pending.take() {
+            if let Some(evicted) = self.cache.add(public_key, share) {
+                trace!("evicted old key from cache: {:?}", evicted);
+            }
+            trace!("finalised DKG: {:?}", public_key);
+        }
+    }
+}
+
+/// Implementation of simple cache.
+#[derive(Debug)]
+pub struct KeyCache {
+    map: BTreeMap<Vec<u8>, SectionKeyShare>,
```

Comment:
The cache is going to have only a few entries in practice (less that 100 I imagine?). So maybe we don't even need this map at all as simple linear lookup through the `list` should be efficient enough (and due to cache locality might be even more efficient than the map lookup, if the number of entries is small).

Reply:
Yep, agree.